### PR TITLE
Download is hanging when one node is not responsive

### PR DESCRIPTION
**Pull request by mniewrzal**
What: Change closes download connection in parallel and sets a timeout for communication between uplink and storage node (uplink side)

Why: Currently downloading even small (10MB) file can hang if one of storage node is not responding. Download operation hangs forever and the process can be only terminated. 

https://storjlabs.atlassian.net/browse/V3-1547

Thanks for submitting a PR!

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Could the PR be broken into smaller PRs?
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?

### CODE REVIEW

File: pkg/eestream/decode.go [hunk 1]
```diff
@@ -101,18 +101,24 @@ func (dr *decodedReader) Close() error {
 	// avoid double close of readers
 	errorThreshold := len(dr.readers) - dr.scheme.RequiredCount()
 	dr.close.Do(func() {
-		var errlist errs.Group
-		// close the readers
+		var errlist errgroup.Group
```
Comment (egonelbre) [inline pkg/eestream/decode.go]: `errgroup` returns only the first error instead of all of them

Comment (mniewrzal) [reply pkg/eestream/decode.go]: do we have something that will fit here or I just need to use for example `sync.WaitGroup`?

Comment (egonelbre) [reply pkg/eestream/decode.go]: As far as I know, no, but might makes sense to implement one.

**Comment by egonelbre** [comment]
Add "Why" to the issue and test for it.

**Comment by mniewrzal** [comment]
Ok, in a moment. Its still wip and mainly for testing.

**Comment by mniewrzal** [comment]
I think it can be reviewed but still looking for a way to test it.

File: pkg/eestream/decode.go [hunk 2]
```diff
@@ -102,16 +102,29 @@ func (dr *decodedReader) Close() error {
 	errorThreshold := len(dr.readers) - dr.scheme.RequiredCount()
 	dr.close.Do(func() {
 		var errlist errs.Group
```
Comment (egonelbre) [inline pkg/eestream/decode.go]: Let's make a nice type for it in `errs2.` e.g. `errs2.WaitGroup` or something.

Comment (mniewrzal) [reply pkg/eestream/decode.go]: added `errs2.Group`

File: internal/errs2/group.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+// Go calls the given function in a new goroutine.
+func (group *Group) Go(f func() error) {
+	group.wg.Add(1)
+
+	go func() {
+		defer group.wg.Done()
+
+		if err := f(); err != nil {
+			group.errOnce.Do(func() {
+				group.errors = append(group.errors, err)
```
Comment (egonelbre) [inline internal/errs2/group.go]: This returns only one error.

File: internal/errs2/group.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+// Go calls the given function in a new goroutine.
+func (group *Group) Go(f func() error) {
+	group.wg.Add(1)
+
+	go func() {
+		defer group.wg.Done()
+
+		if err := f(); err != nil {
+			group.errors = append(group.errors, err)
```
Comment (egonelbre) [inline internal/errs2/group.go]: this is a data race

Comment (mniewrzal) [reply internal/errs2/group.go]: ah, sorry, I should focus on one thing at once

File: pkg/transport/timeout.go [hunk 5]
```diff
@@ -22,3 +23,71 @@ func (it InvokeTimeout) Intercept(ctx context.Context, method string, req interf
...
+	})
+}
+
+func (wrapper *clientStreamWrapper) RecvMsg(m interface{}) error {
+	return wrapper.addTimeout(func() error {
+		return wrapper.stream.RecvMsg(m)
+	})
+}
+
+func (wrapper *clientStreamWrapper) addTimeout(f func() error) error {
```
Comment (egonelbre) [inline pkg/transport/timeout.go]: `withTimeout`

File: internal/errs2/group.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+import "sync"
+
+// Group is a collection of goroutines working on subtasks that are part of
+// the same overall task.
+type Group struct {
+	wg     sync.WaitGroup
+	errors []error
+
+	mutex sync.Mutex
```
Comment (egonelbre) [inline internal/errs2/group.go]: Order these such it can be understood what is being protected and how.

Comment (egonelbre) [reply internal/errs2/group.go]:
```
group  sync.WaitGroup
mu     sync.Mutex
errors []error
```

File: pkg/eestream/decode.go [hunk 7]
```diff
@@ -98,21 +99,25 @@ func (dr *decodedReader) Read(p []byte) (n int, err error) {
 func (dr *decodedReader) Close() error {
 	// cancel the context to terminate reader goroutines
 	dr.cancel()
-	// avoid double close of readers
 	errorThreshold := len(dr.readers) - dr.scheme.RequiredCount()
+	closeGroup := &errs2.Group{}
```
Comment (egonelbre) [inline pkg/eestream/decode.go]: No need for the pointer, use `var closeGroup errs2.Group`

File: pkg/eestream/decode.go [hunk 8]
```diff
@@ -98,21 +99,25 @@ func (dr *decodedReader) Read(p []byte) (n int, err error) {
...
 	dr.close.Do(func() {
-		var errlist errs.Group
-		// close the readers
 		for _, r := range dr.readers {
-			err := r.Close()
-			if err != nil {
-				errlist.Add(err)
-				errorThreshold--
-			}
+			r := r
```
Comment (jtolds) [inline pkg/eestream/decode.go]: this is such a fun go footgun. might be worth leaving a comment how necessary this line is and what subtle bugs will befall you if you remove it

File: pkg/eestream/decode.go [hunk 9]
```diff
@@ -98,21 +99,25 @@ func (dr *decodedReader) Read(p []byte) (n int, err error) {
...
+			closeGroup.Go(func() error {
+				return r.Close()
+			})
 		}
+
 		// close the stripe reader
-		errlist.Add(dr.stripeReader.Close())
-		dr.closeErr = errlist.Err()
+		closeGroup.Go(func() error {
```
Comment (jtolds) [inline pkg/eestream/decode.go]: can you just do `closeGroup.Go(dr.stripeReader.Close)`?

File: pkg/transport/timeout.go [hunk 10]
```diff
@@ -22,3 +23,71 @@ func (it InvokeTimeout) Intercept(ctx context.Context, method string, req interf
...
+
+func (wrapper *clientStreamWrapper) RecvMsg(m interface{}) error {
+	return wrapper.withTimeout(func() error {
+		return wrapper.stream.RecvMsg(m)
+	})
+}
+
+func (wrapper *clientStreamWrapper) withTimeout(f func() error) error {
+	timoutTicker := time.NewTicker(wrapper.timeout)
+	defer timoutTicker.Stop()
```
Comment (jtolds) [inline pkg/transport/timeout.go]:
```
doneCh := make(chan struct{})
defer close(doneCh)
```

File: pkg/transport/timeout.go [hunk 11]
```diff
@@ -22,3 +23,71 @@ func (it InvokeTimeout) Intercept(ctx context.Context, method string, req interf
...
+
+	go func() {
+		select {
+		case <-timoutTicker.C:
+			wrapper.cancel()
+		case <-wrapper.Context().Done():
```
Comment (jtolds) [inline pkg/transport/timeout.go]:
```case <-doneCh:
```

File: pkg/transport/timeout.go (hunk 10 above)
Comment (jtolds) [reply pkg/transport/timeout.go]: this makes it so we can clean up the goroutine early if f returns early

**Pull request merged**